### ISSUE TITLE
[ban-types] Should not ban `{}` by default!

### ISSUE THREAD

**Issue by Mister-Hope** [typescript-eslint/typescript-eslint#2063]
A  lot of type declaration is using `Type Aliases`, to let user fill in.

For example a interface like this: (I got it from an official repo)

```ts
  interface Custom<
    Detail = Record<string, any>,
    TargetDataset = Record<string, any>,
    CurrentTargetDataset = Record<string, any>,
    Mark = Record<string, any>
  > extends Base<TargetDataset, CurrentTargetDataset, Mark> {
    detail: Detail;
  }
```

Sometimes detail is just an empty object! So it's good for me to write `Custom<{}>`, but V3 will throw an error ( not even a warning)

```
Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead.
```

In order to stop this error, I have to write  `Custom<Record<never,never>>`, which is unneeded and a waste of time.

So stop baning `{}` by default. May packages are holding certain objects for certain data, and to prevent the `Cannot read property 'xxx' of undefined`, they are usually initialize with `{}`, so If I know that those fields are not containing any data, I should use `{}`

**Comment by TrejGun** [comment]
its a commion to use `React.ChangeEvent<{}>` or `Promise<{}>` when result is not expectred

**Comment by bbarry** [comment]
It isn't banning the value `{}`

It is banning the type.

As the error states 

     `{}` actually means "any non-nullish value".

That means for your interface if it had `{}` as a type would permit almost any value in that position:

https://www.typescriptlang.org/play/index.html?ssl=29&ssc=1&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgEJwM4QDwBU5QDmEYAInGJiQDTIDCArlFBOPkSeZVmLQLIEA1gD5kAb2QBfAFDTQkWIhSMMYAPYBbbNOS7kpEnGAAbZAF5kAJQgI1UACbZVUUIVpwQAT2HUde9sRkFFRg5lY2do7Oru5ePn66jMysYAGcwTxh1rYOTmAuIG7IHt6+esgCUIJZEbnRhbHefqIQAB6QIPYYaFR4BIFcIbRJLGz96dw0FUKiYgnI9oYmAFz6S8YA3NIy8tDwSPQMqprojvMGlCZhYpJl-uNBk6EWN3eJTKOpD4OZL7fzlWqf2ayDaHS6PSwfQ4jyGh2SYxhPymgNm80Wl2MqwuRk221kxhIyAQAEZVip1Bprgt1qsblItgB6RnIAACYAwAFo2gAHGxgbnMOzSUnUjG41YAckw9hgkoZ0mZbI53NafIQAugUGFopeNMxqxJCukhNCACM4PZyUdKacxbTxJIFRb7PaDchpRhZfLJFsXW6Jcgjb7pEA



As to it being common... doesn't mean it is correct.

**Comment by Mister-Hope** [comment]
> It isn't banning the value `{}`
> 
> It is banning the type.
> 
> As the error states
> 
> ```
>  `{}` actually means "any non-nullish value".
> ```
> 
> That means for your interface if it had `{}` as a type would permit almost any value in that position:
> 
> https://www.typescriptlang.org/play/index.html?ssl=29&ssc=1&pln=1&pc=1#code/JYOwLgpgTgZghgYwgAgEJwM4QDwBU5QDmEYAInGJiQDTIDCArlFBOPkSeZVmLQLIEA1gD5kAb2QBfAFDTQkWIhSMMYAPYBbbNOS7kpEnGAAbZAF5kAJQgI1UACbZVUUIVpwQAT2HUde9sRkFFRg5lY2do7Oru5ePn66jMysYAGcwTxh1rYOTmAuIG7IHt6+esgCUIJZEbnRhbHefqIQAB6QIPYYaFR4BIFcIbRJLGz96dw0FUKiYgnI9oYmAFz6S8YA3NIy8tDwSPQMqprojvMGlCZhYpJl-uNBk6EWN3eJTKOpD4OZL7fzlWqf2ayDaHS6PSwfQ4jyGh2SYxhPymgNm80Wl2MqwuRk221kxhIyAQAEZVip1Bprgt1qsblItgB6RnIAACYAwAFo2gAHGxgbnMOzSUnUjG41YAckw9hgkoZ0mZbI53NafIQAugUGFopeNMxqxJCukhNCACM4PZyUdKacxbTxJIFRb7PaDchpRhZfLJFsXW6Jcgjb7pEA
> 
> As to it being common... doesn't mean it is correct.

Yes, that's it! There is and should be an 

# an empty object

 at that position!

## Please check out carefully what others wants to say before replying to them ok?
@bbarry

**Comment by Mister-Hope** [comment]
Here is an example here

```ts
interface Sample<T = Record<string,any>> {
  aaa: {};
}
```

Mant repos are writing this, becase there are some people who do not want a strict type check, they would like to write simply with out throwing errors, so they can write:

```ts
const a: AAA = { aaa: { a: 1, b: 2} };
```

like this, and make `aaa` as a `Record<string,any>` by default. So it's safe for them to write like:

```ts
console.log(a.aaa.a);

a.aaa.c = 3
```

While others wound want a strict type check, so they write:

```ts
const a: AAA<{ a: number, b?:number}> = { aaa: { a: 1} };
```

And ts can throw errors if they write like:

```ts
a.aaa.c = 1;
a.aaa.b= false;

a.aaa.a.split(' ');
```

But sometimes, we just want that property to be an **empty object**! We wound like typescript throwing errors when we write:

```
a.aaa.a =1;

console.log(a.aaa.b);
```

So we are writing:

```ts
const a:AAA<{}> = { aaa: {} };
```

Which the plugin now throws an error, and we have to write:

```ts
const a:AAA<Record<never, never>> = { aaa: {} };
```

That's definately not reasonable! **Am I clear?** @bbarry

**Comment by bbarry** [comment]
No you are not clear because that code is not reasonable:

    interface Sample<T = Record<string,any>> {
      aaa: {};
    }

The generic type here is unused. The type definition is almost certainly not what you actually think you mean.

I'll assume you meant:

    interface AAA<T = Record<string,any>> {
      aaa: T;
    }

which means that these are both fine:

    const a: AAA = { aaa: { a: 1, b: 2} };
    const a: AAA<{ a: number, b?:number}> = { aaa: { a: 1} };


This however:

    const a:AAA<{}> = { aaa: {} };

does NOT mean "we just want that property to be an empty object."

instead it means "that property can be almost anything"

This:

    const a:AAA<Record<never, never>> = { aaa: {} };

also doesn't mean "we just want that property to be an empty object." Instead it means "we want an object and aren't going to use any members that happen to be there."

If you really mean "we just want that property to be an empty object" then you need some fancy types:

    type NoElements<T> = { [P in keyof T]: never };
    type EmptyObject = NoElements<Record<string,never>>;

    const e: AAA<EmptyObject> = { aaa: {} };
    // @ts-expect-error
    e.aaa = { asdf: 1 };


[Playground link](https://www.typescriptlang.org/play/index.html?target=7&ssl=32&ssc=21&pln=26&pc=1#code/JYOwLgpgTgZghgYwgAgILoDwBVkF5kBKECA9lACYYDOYUoA5gDRwgCeAfO8gN4BQyyOEIBcyLAG5eAX169SIGoNHpUeHoJHq4ogIyNkAI1EAmZFLOS5JBWEPLM3JchABXALYHo+gwH5Rrj2gzLnxHIW0tXTMLWSsbZAR7VAxuKRCtTVSYuQA6cLUdSQQ8oTUABkkAekrkAAEwKgBaCAAPAAdiMGaoKDJkAAoANzgAGxcUEhhkMFYOnnMEFhASW09BKipgehAIcmcXEZGASlz8-FdDqpr6ptaOhC7oXqgB4bGJqZm5rMWQZdWUHANlsdnsXCByBAYKBdidimdkODIdDQUUSnA1AByTGWOKKchJDBEUgUDA7QZeZwQClQThqMKZcxSSzVZAjEgkADWVDZwE5KGJZEo5K8ItpyGAPIA7r0QPRkDA+gAiCBuNozZAkAwAK06SuQbggLAYvHI6PpTh02V4XxQADkSABREaqiDgKjYdKOADaAAUJSBkPzWJMxABdfzUoLMm2zFCOtUzADyOs6agdztd7qJxCF1FoDEYYs4lnkigghIT6tYKd1Dy9GgiWRjrJuzXanW6z14EHNoXW5BgUWZQA)

**Comment by Mister-Hope** [comment]
Excuse me, from which docs do you think `{}` means "that property can be almost anything"

Then why am I catching problems writing:

![image](https://user-images.githubusercontent.com/33315834/82686513-8a6b6200-9c88-11ea-9142-ffc481293bbb.png)

**Comment by Mister-Hope** [comment]
Excuse me, from which docs do you think `{}` means "that property can be almost anything"

Then why am I catching problems writing:

![image](https://user-images.githubusercontent.com/33315834/82686513-8a6b6200-9c88-11ea-9142-ffc481293bbb.png)

**Comment by bbarry** [comment]
Because the `{}` type does restrict what you can do with the type (it pretends to not know about any members that the type has for example). It just doesn't restrict what values are instances of `{}` (except `null` or `undefined`).

```typescript
interface AAA<T = Record<string,any>> {
  aaa: T;
}

const a: AAA = { aaa: { a: 1, b: 2 } };
const b: AAA<{ a?: number, b?: number }> = { aaa: { a: 1 } };
const c: AAA<{}> = { aaa: {} };
const d: AAA<Record<never, never>> = { aaa: {} };
type NoElements<T> = { [P in keyof T]: never };
type EmptyObject = NoElements<Record<string,never>>;
const e: AAA<EmptyObject> = { aaa: {} };

a.aaa.error = 1;
b.aaa.error = 1; // Property 'error' does not exist on type '{}'.(2339) 
c.aaa.error = 1; // Property 'error' does not exist on type '{}'.(2339) 
d.aaa.error = 1; // Property 'error' does not exist on type 'Record<never, never>'.(2339)
e.aaa.error = 1; // Type '1' is not assignable to type 'never'.(2322)

a.aaa = {};
b.aaa = {};
c.aaa = {};
d.aaa = {};
e.aaa = {};

a.aaa = { a: 1 };
b.aaa = { a: 1 };
c.aaa = { a: 1 };
d.aaa = { a: 1 };
e.aaa = { a: 1 }; // Type 'number' is not assignable to type 'never'.(2322)

a.aaa = { error: 1 };
b.aaa = { error: 1 }; // Type '{ error: number; }' is not assignable to type '{ a?: number | undefined; b?: number | undefined; }'.(2322)
c.aaa = { error: 1 };
d.aaa = { error: 1 };
e.aaa = { error: 1 }; // Type 'number' is not assignable to type 'never'.(2322)

a.aaa = 1; // Type '1' is not assignable to type 'Record<string, any>'.(2322)
b.aaa = 1; // Type '1' has no properties in common with type '{ a?: number | undefined; b?: number | undefined; }'.(2559)
c.aaa = 1;
d.aaa = 1;
e.aaa = 1; // Type '1' is not assignable to type 'NoElements<Record<string, never>>'.(2322)
```

**Comment by bradzacher** [comment]
This is actually the exact reason that the rule now bans the `{}` ***type***.
It's a common misconception that the `{}` ***type*** is the same as the `{}` ***value***.

But as the error message states: _this is not the case_!

The ***type*** `{}` doesn't mean "any empty object", it means "any non-nullish value".

This is obviously a huge type safety hole!

Taking your example, the following code is _completely typechecked valid_, even though it makes no sense to be:

```ts
interface AAA {
  aaa: {};
}

const x: AAA = { aaa: true };
```
[repl](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgILuQbwFDOXAgLiwF8BubE7bBAexAGcxkAPY9VZAXi3yOTBQArinLYgA)

Unfortunately, there's no _type_ in TS that means "an empty object".
But at the very least if you instead use `Record<string, unknown>`, then you can enforce that it's an object, and you can enforce that the consumer also checks the type of the key before using it:

```ts
interface AAA {
  aaa: {[k: string]: unknown};
}

const x: AAA = { aaa: {} };

const test = x.aaa.a;
test.foo; // test is unknown, so error
```

**Comment by bbarry** [comment]
I am pretty confident these are all valid definitions for empty object:

```typescript

type NoElements<T> = { [P in keyof T]: never };
type EmptyObject1 = NoElements<Record<string, never>>;

type EmptyObject2 = Omit<Record<string, never>, keyof string>;

type EmptyObject3 = Omit<Record<any, never>, keyof any>;
```

both `Omit` and `never` are very powerful language constructs for blacklisting things.

https://www.typescriptlang.org/play/index.html?target=7&ssl=7&ssc=57&pln=1&pc=1#code/FAFwngDgpgBAcgewKIBsoFsoDsQGcA8AKgHwwC8MA3jANoAKMAlljANZRgIBmMhAugC4YWKADcoAJxgBfANyhIsJOgjgA8gCMAVlADGIAIzl4yNJhwEASnoQSAJvlwgJzAOYAaYWMnFi8hdAwyqpgmjr6AEzGauiMIPjWurYOTi5YHl7iEsSe7Jw8qW5+wAFKKuraeiAAzNGx8YnJ+ACGWGCeIlk5bBzcMK1gxcBoIP1CwRXhNf7NxlgArigo8rMU81h2UFzMUHYrxgAM+xQGxzBc6-qMCCwAFACUVDJnYVUAdBASCCDfimciAHcYK99A8zgBycFnADKYHQGgQKDBwFWTwAHkIDs8SqjKHJgEA

**Comment by mgcrea** [comment]
Just been hit by this, in my case with React empty Props:

```tsx
type Props = {};

const Button: FunctionComponent<Props> = () => {
  // ...
}
```

Is a pretty common pattern especially when you rely on user snippets to speed up development.

I understand that the type safety is lost, but IMO it is a typescript bug (`{}` should be the type of an empty object) and enforcing this eslint rule will lead to a loss of both productivity (ability to quickly add props to an empty object) and readability.

What's frustrating is that it looks like we can't really disable this only ban-type without having to inline the whole rule config, messy! Guess I should open a separate issue about that.

**Comment by bbarry** [comment]
It doesn't matter how common it is. That doesn't make it a good practice:

https://github.com/microsoft/TypeScript/wiki/FAQ#why-are-all-types-assignable-to-empty-interfaces

>  ### Why are all types assignable to empty interfaces?
> > I wrote some code like this and expected an error:
> > ```ts
> > interface Thing { /* nothing here */ }
> > function doSomething(a: Thing) {
> >   // mysterious implementation here
> > }
> > // Expected some or all of these to be errors
> > doSomething(window);
> > doSomething(42);
> > doSomething('huh?');
> > ```
> Types with no members can be substituted by *any* type.
> In this example, `window`, `42`, and `'huh?'` all have the required members of a `Thing` (there are none).
> In general, you should *never* find yourself declaring an `interface` with no properties.

The crazy omit type is specifically an empty object. The only valid value a variable can be assigned if it is that type is the value `{}`.

**Comment by zetoke** [comment]
@bbarry what's about 
```typescript
interface IState {
  field: string;
}

class MyComponent extends React.Component<{}, IState> {
...
}
```

the default value for props and state is `{}` in React components. And it makes sense since I can catch something like this:
```
public render() {
  const a = this.props.kek;
  console.log(a);
}
```
`Property 'kek' does not exist on type 'Readonly<{}> & Readonly<{ children?: ReactNode; }>'.ts(2339)`

And this error is just muted with `Record<string, unknown>`

**Comment by nstepien** [comment]
Yes I also feel `{}` shouldn't be banned from being used as a generic, for example
```ts
React.PropsWithChildren<{}>
```
where `PropsWithChildren` is
```ts
type PropsWithChildren<P> = P & { children?: ReactNode };
```
In that case we don't want to allow any index signature.

**Comment by bbarry** [comment]
Can't you use unknown for that?

```typescript
interface IState {
  field: string;
}

class MyComponent extends React.Component<unknown, IState> {
  public foo() {
    // @ts-expect-error Property 'kek' does not exist on type 'Readonly<unknown> & Readonly<{ children?: ReactNode; }>'.(2339)
    const a = this.props.kek;
    console.log(a);
  }
}
```


If you intend on adding props later and want to track all of these unknowns with a named unknown type like `type MyProps = unknown;`

If you really wanted you could declare MyProps like this:

```typescript
const __id = Symbol();
export class MyProps {
  [__id]: Symbol;

  constructor() { this[__id] = __id; }
};

export function isMyProps(input: unknown): input is MyProps {
    return (input as MyProps)?.[__id] === __id;
}
```
```

**Comment by nstepien** [comment]
`unknown` does work indeed, thanks. It's not very intuitive, I wish TS would address `{}` somehow.

**Comment by bbarry** [comment]
I don't think typescript can fix it with `{}` or any empty interface type without going away from structural typing in general.

You could also try using an explicit empty props object:

```typescript
type MyProps = Omit<Record<string, never>, keyof string>;
```

This explicitly tells consumers that your props object cannot have any members:

```typescript

type MyProps = Omit<Record<string, never>, keyof string>;

interface IState {
  field: string;
}

class MyComponent extends React.Component<MyProps, IState> {
  public foo() {
    // no error unfortunately but you could lint this probably with a lint rule that looks for variables of this type
    const a = this.props.kek; // type: never 
    console.log(a);
  }
}

**Comment by phiresky** [comment]
It would be great if a suggestion could be added to the rule for what to use when you use {} as meaning "empty object". Most uses that I have are where the type is `&` with some other type, and I just want it to not change the other type.

`unknown` does work for this, but it's not intuitive at all. `Record<string, never>` also works. `Omit<Record<string, never>, keyof string>` works and is more strict. One of these should be added as a third suggestion of how to fix the lint.

**Comment by bradzacher** [comment]
Could you give an example of your use case?

It's hard to provide a general recommendation.

It is actually safe to use `{}` in an intersection type, but it's rare you need to do that (as it is a no op)

**Comment by leethree** [comment]
I'm still very confused about what `{}` means, and it seems like almost everybody use it as an object. It's also used extensively in the [typings of `react`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts).

It's working perfectly well for typing React components without props:

```tsx
const Comp = (props: {}) => null; // React component

const a = <Comp /> // Correct

const b = <Comp test={false} /> // Type '{ test: boolean; }' is not assignable to type '{}'. Property 'test' does not exist on type '{}'.(2322)
```

What alternative should I use to replace `{}` in this case?

**Comment by bbarry** [comment]
It isn't that bad of a problem for React because your call to `React.createComponent` is passing the parameter to an intersection type resulting in the actual parameter being the intersection `React.Attributes & {}`. Unfortunately there is no way for ts-eslint to know that is a React component and the type of the parameter is actually going to be an intersection with some other interface when it actually gets used.


I would consider:

    const Comp = (props: React.Attributes) => null; 


The thing is that you can call that function just like any ordinary function:

    const result = Comp({});

With the type specified like that the following code is valid:

    const result = Comp(4);

which is surely not what you intend.

**Comment by leethree** [comment]
@bbarry Thanks for the explanation. It makes sense to me. I feel TypeScript should clarify and provide a helper type for empty object because it's used a lot.

**Comment by uhoh-itsmaciek** [comment]
I ran into exactly what @zetoke mentioned above: empty props, but non-empty state. I thought the solution you mention [here](https://github.com/typescript-eslint/typescript-eslint/issues/2063#issuecomment-634666407) would work, @bbarry, but playing with it in the TS playground (3.9.2), I noticed it will actually let you reference any properties in the function body (these are typed `never`, so there's little potential for misuse, but apparently I can still pass use those as, e.g., arguments to functions like `console.log` that take `any` arguments).

Here are the different things I tried:

```
// the violation at the heart of the discussion
type Props1 = {}

const c2 = (p: Props1) => {
    console.log(p.doesNotExist) // TS complains 👍
}

c2(42) // TS does not complain 😮
c2({})

// proposed above
type Props2 = Omit<Record<string, never>, keyof string>

const c1 = (p: Props2) => {
    console.log(p.doesNotExist) // TS does not complain 😮
}

c1(42) // TS complains 👍
c1({})

// using unknown
type Props3 = unknown

const c3 = (p: Props3) => {
    console.log(p.doesNotExist) // TS complains 👍
}

c3(42) // TS does not complain 😮
c3({})
```

(I'm using a simple function instead of a React component because it's simpler, but I think this translates to the class-based component discussion pretty easily). I think there are two things we want to ensure when declaring a type on a function: that callers are prevented from passing in bogus values, and that the specified type represents the value within the function body to avoid implementation mistakes. I think the interesting thing here is that you're almost never calling a React component function (or creating a React component class instance) directly, so the first part is largely moot (in the no-props case). This is more about saying "I don't want the type of the props object to have any properties I (or a future maintainer) might accidentally reference in the component implementation". Given that, I think `unknown` is probably more useful for React (even though it has the same exact issue the rule appears to be warning against).

It'd be nice to have more explicit guidance in the hint for the React case, though. Right now, it says

```
  6:3  error  Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead  @typescript-eslint/ban-types
```

I don't really want either of these: I want a type meaning "an object with no properties". Do you think it's reasonable to expand this text to include a hint for that case?

**Comment by uhoh-itsmaciek** [comment]
(incidentally, React defines Attributes as

```ts
    interface Attributes {
        key?: Key;
    }
```

so this is probably not a great alternative)

**Comment by misuken-now** [comment]
For reference, there was a way to complain both inside and outside the function.

```
type Props = {}

const c1 = <V extends Props>(p: V & (V extends number | string | boolean | Function | unknown[] ? never : Props)) => {
    console.log(p.doesNotExist) // TS complains 👍
}

c1(42); // TS complains 👍
c1("foo"); // TS complains 👍
c1(true); // TS complains 👍
c1(() => undefined); // TS complains 👍
c1([]); // TS complains 👍
c1({});
```

**Comment by bradzacher** [comment]
We will not be removing `{}` from the rule defaults, as it is an unsafe type.

I have updated [my comment above](https://github.com/typescript-eslint/typescript-eslint/issues/2063#issuecomment-632833366) with more information and a workaround for react users.
